content/posts/2022-05-13-calculating-type-sets.markdown

```diff
@@ -0,0 +1,498 @@
+---
+
+Go 1.18 added the biggest and probably one of the most requested features of
+all time to the language: [Generics](https://go.dev/doc/go1.18#generics). If
+you want a comprehensive introduction to the topic, there are many out there
+and I would personally [recommend this talk I gave at the Frankfurt Gopher
+Meetup](https://www.youtube.com/watch?v=QP6v-Q5Foek).
+
+This blog post is not an introduction to generics, though. It is about [this
+sentence from the spec](https://go.dev/ref/spec#Operands):
+
+> Implementation restriction: A compiler need not report an error if an
+> operand's type is a type parameter with an empty type set.
+
+This restriction might seem strange to you. After all, if a type set is empty,
+it seems very helpful to report that to the user, because they obviously made a
+mistake - an empty type set can never be used as a constraint, as the function
+using it could never be instantiated.
+
+I want to explain the reason that sentence is there and also go into a couple
+of related design decisions of the generics design. I'm trying to be expansive
+in my explanation, which means that you should not *need* any special knowledge
+to understand it. It also means, some of the information might be boring to
+you - feel free to skip the corresponding sections.
+
+The reason that sentence is in the Go spec, is that it turns out to be hard to
+actually determine if a type set is empty. Hard enough, that the Go team did
+not want to require an implementation to solve that. Let's see why.
+
+## P vs. NP
+
+When we talk about whether or not a problem is hard, we often group problems
+into two big classes:
+
+1. Problems which can be *solved* reasonably efficiently. This class is called
+   P.
+2. Problems which can be *decided* reasonably efficiently. This class is called
+   NP.
+
+The first obvious follow up question is “what does ‘reasonably efficient’
+mean?”. The answer to that is “there is an algorithm with a running time
+polynomial in its input size”[^1].
+
+The second obvious follow up question is “what's the difference between
+‘solving’ and ‘deciding’?”.
+
+*Solving* a problem means what you think it means: Finding a solution. If I
+give you a number and ask you to solve the factorization problem, I'm asking
+you to find a (non-trivial) factor of that number.
+
+*Deciding* a problem means that I give you solution and I'm asking you if the
+solution is correct. For the factorization problem, I'd give you two numbers
+and ask you to verify that the second is a factor of the first.
+
+These two things are often very different in difficulty. If I ask you to give
+me a factor of 297863737, you probably know no better way than to sit down and
+try to divide it by a lot of numbers and see if they divide it evenly. But if I
+ask you to verify that 9883 is a factor of that number, you just have to do a
+bit of long division and it either divides it, or it does not.
```

Comment:
nice explanation of P vs NP. I wish they had done it this way at uni.

Reply:
Thanks :) TBF, it's just an informal explanation, so I'm not bogged down by formalism. Then again, nothing would prevent professors from giving *both* an informal explanation *and* a formal proof - and the ones I always liked best tended to do that.